compiler/docgen.nim

```diff
@@ -1262,18 +1278,26 @@ proc genSection(d: PDoc, kind: TSymKind, groupedToc = false) =
       "sectionid", $ord(kind), "sectionTitle", title,
       "sectionTitleID", $(ord(kind) + 50), "content", d.section[kind].finalMarkup]
 
-  var tocSource = d.toc
+  proc cmp(x, y: TocItem): int = cmpIgnoreCase(x.sortName, y.sortName)
   if groupedToc:
-    for p in d.tocTable[kind].keys:
+    let overloadableNames = toSeq(keys(d.tocTable[kind]))
+    for plainName in overloadableNames.sorted():
+      var overloadChoices = d.tocTable[kind][plainName]
+      overloadChoices.sort(cmp)
```

Comment:
This looks quite slow, can we do something about it?

Reply:
no it does not look quite slow, O(n lg n) where n is nb of symbols will be dwarfed by other costs (parsing, semcheck, etc); although this could be optimized, there is no point in premature optimization at expense of simplicity.

and indeed:
`nim doc --project --doccmd:skip lib/pure/times.nim` takes 1.5 seconds
the cost for this block of the whole block in `if groupedToc:` is 0.0035, ie 1/445 of total cost for a total of 1936 overloads. There are many optimizations worth considering before considering this one.
```
(0.0035069999999999824, 0.0000020000000000575113, 1936, 1, "D20210723T093627.total")
```

```
diff --git a/compiler/docgen.nim b/compiler/docgen.nim
index 842f95ce7..dde9ba0bb 100644
--- a/compiler/docgen.nim
+++ b/compiler/docgen.nim
@@ -1266,7 +1266,7 @@ proc generateTags*(d: PDoc, n: PNode, r: var string) =
     if not checkForFalse(n[0][0]):
       generateTags(d, lastSon(n[0]), r)
   else: discard
-
+import std/times except getDateStr, getClockStr
 proc genSection(d: PDoc, kind: TSymKind, groupedToc = false) =
   const sectionNames: array[skModule..skField, string] = [
     "Imports", "Types", "Vars", "Lets", "Consts", "Vars", "Procs", "Funcs",
@@ -1280,10 +1280,16 @@ proc genSection(d: PDoc, kind: TSymKind, groupedToc = false) =

   proc cmp(x, y: TocItem): int = cmpIgnoreCase(x.sortName, y.sortName)
   if groupedToc:
+    var dt0 {.global.}: typeof(cpuTime() - cpuTime())
+    var tot0 {.global.}: int
+    let t = cpuTime()
+    var tot = 0
     let overloadableNames = toSeq(keys(d.tocTable[kind]))
     for plainName in overloadableNames.sorted():
       var overloadChoices = d.tocTable[kind][plainName]
+      # echo "D20210723T093748: ", $overloadChoices.len, " " , plainName
       overloadChoices.sort(cmp)
+      tot += overloadChoices.len
       var content: string
       for item in overloadChoices:
         content.add item.content
@@ -1291,6 +1297,11 @@ proc genSection(d: PDoc, kind: TSymKind, groupedToc = false) =
           "sectionid", $ord(kind), "sectionTitle", title,
           "sectionTitleID", $(ord(kind) + 50),
           "content", content, "plainName", plainName]
+    let t2 = cpuTime()
+    let dt = t2 - t
+    dt0 += dt
+    tot0 += tot
+    echo (dt0, dt, tot0, tot, "D20210723T093627.total")
   else:
     for item in d.tocSimple[kind].sorted(cmp):
       d.toc2[kind].add item.content
```

Reply:
> There are many optimizations worth considering before considering this one.

I hoped somebody would be prove me wrong by the numbers, however: 

1. Code is easier to optimize when it's still fresh in the developer's head.
2. Things add up. Today's compiler is significantly slower than a couple of years ago. And the things that made it slower are hard to optimize. At the same time, I still get no help whatsoever in the one architectural change (= IC) that would imply we can stop caring about performance in most of the compiler's code.

Reply:
I agree with `1.`, but I'm skeptical of `2.`; big-O notation and Amdahl's law is what matters in the end and if the compiler is slower today than years ago, it's more likely to be caused by a small number of bottlenecks (that should be identified and fixed) than to a multitude of small things adding up (eg better error messages, static analysis etc).

On the same vein, optimizing PNode (eg removing comment node) will have a bigger impact on memory/performance than optimizing PSym since much more PNode's are allocated.

As for IC, I'm still convinced this is the wrong approach, and `cling` shows that IC can be be done with a compiler as a service (CAS) approach even for a language like C++. Even if nim's current IC approach would end up being faster than current nim full re-compilation (which isn't the case today), the overheads involved would still make CAS much faster. Not only that, but you'd also get a REPL for free in the process, unlike with IC.

Reply:
"compiler as a service (CAS) approach" is inherently non-scalable wrt memory consumption. It's also completely not required, C# and Delphi (Delphi in 1990 btw) have proven beyond doubt that "save to disk" is fast enough for everything, REPL included. Plus IC forces us to eliminate phase-ordering bugs which plague the Nim compiler.